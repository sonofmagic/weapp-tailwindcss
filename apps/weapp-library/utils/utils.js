import Promise from './es6-promise'

/**
 * 函数节流 函数连续调用时，func 执行频率限定为 次 / wait
 * underscore.js 实现
 * @param {Function} func 传入函数
 * @param {number} wait 表示时间窗口的间隔
 * @param {object} options 如果想忽略开始边界上的调用，传入{leading: false}
 * @param {object} options.leading 是否在开始边界上调用，默认true
 * @param {object} options.trailing 是否在结尾边界上调用，默认true
 * @return {Function} 返回客户调用函数
 */
export function throttle(func, wait, options) {
  let context, args, result
  let timeout = null
  // 上次执行时间点
  let previous = 0
  if (!options) { options = {} }
  // 延迟执行函数
  const later = function () {
    // 若设定了开始边界不执行选项，上次执行时间始终为0
    previous = options.leading === false ? 0 : new Date()
    timeout = null
    result = func.apply(context, args)
    if (!timeout) { context = args = null }
  }
  return function () {
    const now = new Date()
    // 首次执行时，如果设定了开始边界不执行选项，将上次执行时间设定为当前时间。
    if (!previous && options.leading === false) { previous = now }
    // 延迟执行时间间隔
    const remaining = wait - (now - previous)
    context = this
    args = arguments
    // 延迟时间间隔remaining小于等于0，表示上次执行至此所间隔时间已经超过一个时间窗口
    // remaining大于时间窗口wait，表示客户端系统时间被调整过
    if (remaining <= 0 || remaining > wait) {
      clearTimeout(timeout)
      timeout = null
      previous = now
      result = func.apply(context, args)
      if (!timeout) { context = args = null }
    // 如果延迟执行不存在，且没有设定结尾边界不执行选项
    }
    else if (!timeout && options.trailing !== false) {
      timeout = setTimeout(later, remaining)
    }
    return result
  }
}

/**
 * 函数防抖 返回函数连续调用时，空闲时间必须大于或等于 wait，func 才会执行
 * https://www.npmjs.com/package/debounce-promise
 * @param {Function} func 传入函数
 * @param {number} wait 表示时间窗口的间隔
 * @param {object} options 其他参数
 * @param {boolean} options.leading 调用触发于开始边界而不是结束边界
 * @param {boolean} options.accumulate 保留现在接收到的所有参数并以数组的方式传递给传入函数
 * @return {Function} 返回客户调用函数
 */
export function debounce(fn, wait = 0, options = {}) {
  let lastCallAt
  let deferred
  let timer
  let pendingArgs = []
  return function debounced(...args) {
    const currentWait = getWait(wait)
    const currentTime = new Date().getTime()

    const isCold = !lastCallAt || (currentTime - lastCallAt) > currentWait

    lastCallAt = currentTime

    if (isCold && options.leading) {
      return options.accumulate
        ? Promise.resolve(fn.call(this, [args])).then(result => result[0])
        : Promise.resolve(fn.call(this, ...args))
    }

    if (deferred) {
      clearTimeout(timer)
    }
    else {
      deferred = defer()
    }

    pendingArgs.push(args)
    timer = setTimeout(flush.bind(this), currentWait)

    if (options.accumulate) {
      const argsIndex = pendingArgs.length - 1
      return deferred.promise.then(results => results[argsIndex])
    }

    return deferred.promise
  }

  function flush() {
    const thisDeferred = deferred
    clearTimeout(timer)

    Promise.resolve(
      options.accumulate
        ? fn.call(this, pendingArgs)
        : fn.apply(this, pendingArgs[pendingArgs.length - 1]),
    )
      .then(thisDeferred.resolve, thisDeferred.reject)

    pendingArgs = []
    deferred = null
  }
}

function getWait(wait) {
  return (typeof wait === 'function') ? wait() : wait
}

function defer() {
  const deferred = {}
  deferred.promise = new Promise((resolve, reject) => {
    deferred.resolve = resolve
    deferred.reject = reject
  })
  return deferred
}

/**
 * 将日期对象转为日期字符串: '2018-04-22'
 * @param glue {String} 连接符号
 */
export function formatDate(date, glue = '-') {
  const year = date.getFullYear()
  const month = date.getMonth() + 1
  const day = date.getDate()
  return [year, month, day].map((n) => {
    n = n.toString()
    return n[1] ? n : `0${n}`
  }).join(glue)
}

/**
 * 日期加减
 * @param interval {String} 要添加的时间间隔的单位
 *        { y: year, q: quarter, M: month, w: week,
            d: day, h: hour, m: minute, s: second}
 * @param number {Number} 要添加的时间间隔
 * @param date 时间对象
 * @return {Date} 新的时间对象
 * 使用：
 * var now = new Date();
 * var newDate = DateAdd( "d", 5, now);
 */
export function dateAdd(interval, number, date) {
  switch (interval) {
    case 'y': {
      date.setFullYear(date.getFullYear() + number)
      return date
    }
    case 'q': {
      date.setMonth(date.getMonth() + number * 3)
      return date
    }
    case 'M': {
      date.setMonth(date.getMonth() + number)
      return date
    }
    case 'w': {
      date.setDate(date.getDate() + number * 7)
      return date
    }
    case 'd': {
      date.setDate(date.getDate() + number)
      return date
    }
    case 'h': {
      date.setHours(date.getHours() + number)
      return date
    }
    case 'm': {
      date.setMinutes(date.getMinutes() + number)
      return date
    }
    case 's': {
      date.setSeconds(date.getSeconds() + number)
      return date
    }
  }
}
