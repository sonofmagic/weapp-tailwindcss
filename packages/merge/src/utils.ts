import type { ClassValue } from 'clsx'
import type {
  createTailwindMerge as _createTailwindMergeV3,
  extendTailwindMerge as _extendTailwindMergeV3,
  twJoin as _twJoinV3,
  twMerge as _twMergeV3,
} from 'tailwind-merge'
import type {
  createTailwindMerge as _createTailwindMergeV2,
  extendTailwindMerge as _extendTailwindMergeV2,
  twJoin as _twJoinV2,
  twMerge as _twMergeV2,
} from 'tailwind-merge-v2'
import type { CreateOptions } from './types'
import { clsx } from 'clsx'
import { escape } from 'weapp-tailwindcss/escape'

export type TailwindMergeVersion = 2 | 3
// 处理 escape 逻辑
export function noop(x: string) {
  return x
}

export interface CreateFactoryOptions<V extends TailwindMergeVersion> {
  version: V
  twMerge: V extends 3 ? typeof _twMergeV3 : typeof _twMergeV2
  twJoin: V extends 3 ? typeof _twJoinV3 : typeof _twJoinV2
  extendTailwindMerge: V extends 3 ? typeof _extendTailwindMergeV3 : typeof _extendTailwindMergeV2
  createTailwindMerge: V extends 3 ? typeof _createTailwindMergeV3 : typeof _createTailwindMergeV2
}

export function createFactory<V extends TailwindMergeVersion>(opts: CreateFactoryOptions<V>) {
  const {
    twMerge: _twMerge,
    twJoin: _twJoin,
    createTailwindMerge: _createTailwindMerge,
    extendTailwindMerge: _extendTailwindMerge,
  } = opts

  return function create(options?: CreateOptions) {
    const escapeFn = options?.escapeFn ?? escape
    const e = options?.disableEscape ? noop : escapeFn

    function twMerge(...inputs: ClassValue[]): string {
      return e(_twMerge(clsx(...inputs)))
    }

    function twJoin(...inputs: ClassValue[]): string {
      return e(_twJoin(clsx(...inputs)))
    }

    // 版本安全的 extendTailwindMerge
    const extendTailwindMerge = (...args: Parameters<typeof _extendTailwindMerge>) => {
      // @ts-ignore
      const customTwMerge = _extendTailwindMerge(...args)
      return function cn(...inputs: ClassValue[]): string {
        return e(customTwMerge(clsx(...inputs)))
      }
    }

    // 版本安全的 createTailwindMerge
    const createTailwindMerge = (...args: Parameters<typeof _createTailwindMerge>) => {
      // @ts-ignore
      const customTwMerge = _createTailwindMerge(...args)
      return function cn(...inputs: ClassValue[]): string {
        return e(customTwMerge(clsx(...inputs)))
      }
    }

    return {
      twMerge,
      twJoin,
      extendTailwindMerge,
      createTailwindMerge,
    }
  }
}
