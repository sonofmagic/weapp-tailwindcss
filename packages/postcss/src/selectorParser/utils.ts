// 选择器 AST 相关的通用工具，提供组合与替换的辅助函数
import type { Node } from 'postcss-selector-parser'
import type { IStyleHandlerOptions } from '../types'
import psp from 'postcss-selector-parser'

export type ParserTransformOptions = Partial<{
  lossless: boolean
  updateSelector: boolean
}>

// normalizeTransformOptions 确保 parser 在修改选择器时保持必要的副作用
export function normalizeTransformOptions(options?: ParserTransformOptions): ParserTransformOptions {
  return {
    lossless: false,
    updateSelector: true,
    ...options,
  }
}

// mklist 将节点扩展为 "节点 + combinator + 节点副本" 的结构
export function mklist(node: Node): Node[] {
  return [
    node,
    psp.combinator({
      value: '+',
    }),
    node.clone(),
  ]
}

// composeIsPseudoAst 按需把字符串数组转换为 :is(...) 结构的 AST
export function composeIsPseudoAst(strs: string | string[]): Node[] {
  if (typeof strs === 'string') {
    return mklist(psp.tag({
      value: strs,
    }))
  }
  if (strs.length > 1) {
    return mklist(psp.pseudo({
      value: ':is',
      nodes: strs.map(str =>
        psp.tag({
          value: str,
        }),
      ),
    }))
  }
  return mklist(psp.tag({
    value: strs[0],
  }))
}

// 根据配置生成替换子代选择器的 AST，默认使用 view 标签
export function getCombinatorSelectorAst(options: IStyleHandlerOptions) {
  let childCombinatorReplaceValue: Node[] = mklist(psp.tag({ value: 'view' }))
  const { cssChildCombinatorReplaceValue } = options
  if (
    typeof cssChildCombinatorReplaceValue === 'string'
    || (Array.isArray(cssChildCombinatorReplaceValue) && cssChildCombinatorReplaceValue.length > 0)) {
    childCombinatorReplaceValue = composeIsPseudoAst(cssChildCombinatorReplaceValue)
  }
  return childCombinatorReplaceValue
}
