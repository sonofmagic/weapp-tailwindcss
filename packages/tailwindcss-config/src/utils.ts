export {
  defu,
  defuOverrideArray,
} from '@weapp-tailwindcss/shared'
