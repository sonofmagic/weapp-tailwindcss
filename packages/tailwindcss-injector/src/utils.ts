export {
  defu,
  defuOverrideArray,
  isRegexp,
  regExpTest,
} from '@weapp-tailwindcss/shared'

export {
  md5,
  removeAllExtensions as removeFileExtension,
} from '@weapp-tailwindcss/shared/node'
