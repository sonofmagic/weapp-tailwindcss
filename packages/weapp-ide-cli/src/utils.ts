import process from 'node:process'
import path from 'pathe'
import { execa } from 'execa'
import type { AliasEntry } from './types'

export async function execute(cliPath: string, argv: string[]) {
  const task = execa(cliPath, argv)

  task?.stdout?.pipe(process.stdout)

  await task
  // 调用返回码为 0 时代表正常，为 -1 时错误。
  // task.exitCode === 0
  // task.exitCode === -1
}

export function resolvePath(filePath: string) {
  if (path.isAbsolute(filePath)) {
    return filePath
  }
  else {
    return path.resolve(process.cwd(), filePath)
  }
}

function alias(argv: string[], entry: AliasEntry) {
  let findIdx = argv.indexOf(entry.find)
  // alias -p as --project
  if (findIdx > -1) {
    argv[findIdx] = entry.replacement
  }
  else {
    findIdx = argv.indexOf(entry.replacement)
  }

  if (findIdx > -1) {
    const paramIdx = findIdx + 1
    const param = argv[paramIdx]
    // 存在项目目录
    if (param && param[0] !== '-') {
      argv[paramIdx] = resolvePath(param)
    }
    else {
      argv.splice(paramIdx, 0, process.cwd())
    }
  }
  return argv
}

function pathCompat(argv: string[], option: string) {
  const findIdx = argv.indexOf(option)

  if (findIdx > -1) {
    const paramIdx = findIdx + 1
    const param = argv[paramIdx]
    // 存在项目目录
    if (param && param[0] !== '-') {
      argv[paramIdx] = resolvePath(param)
    }
    else {
      argv.splice(paramIdx, 0, process.cwd())
    }
  }
  return argv
}

export function createAlias(entry: AliasEntry) {
  return function (argv: string[]) {
    return alias(argv, entry)
  }
}

export function createPathCompat(option: string) {
  return function (argv: string[]) {
    return pathCompat(argv, option)
  }
}
