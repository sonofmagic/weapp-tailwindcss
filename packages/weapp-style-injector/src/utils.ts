import type { PerFileImportResolver } from './core'

export function toArray<T>(value: T | T[] | null | undefined): Array<NonNullable<T>> {
  if (value == null) {
    return []
  }
  return (Array.isArray(value) ? value : [value]) as Array<NonNullable<T>>
}

export function mergePerFileResolvers(
  resolvers: Array<PerFileImportResolver | null | undefined>,
): PerFileImportResolver | undefined {
  const activeResolvers = resolvers.filter((resolver): resolver is PerFileImportResolver => typeof resolver === 'function')

  if (activeResolvers.length === 0) {
    return undefined
  }

  return (fileName: string) => activeResolvers.flatMap(resolver => toArray(resolver(fileName)))
}

export function ensurePosix(value: string): string {
  return value.replace(/\\/g, '/')
}

export function normalizeRoot(root: string): string {
  const trimmed = root.trim().replace(/^[./\\]+/, '').replace(/\\+/g, '/')
  return trimmed.endsWith('/') ? trimmed.slice(0, -1) : trimmed
}

export function normalizeRelativeImport(target: string): string {
  if (target.startsWith('.') || target.startsWith('/')) {
    return target
  }
  return `./${target}`
}
