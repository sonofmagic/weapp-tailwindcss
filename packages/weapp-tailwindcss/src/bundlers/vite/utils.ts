import type { ExistingRawSourceMap } from 'rollup'
import path from 'node:path'
import process from 'node:process'

export function slash(p: string): string {
  return p.replace(/\\/g, '/')
}

export const isWindows = process.platform === 'win32'

const cssLangs = `\\.(css|less|sass|scss|styl|stylus|pcss|postcss)($|\\?)`
export const cssLangRE = new RegExp(cssLangs)
export function isCSSRequest(request: string): boolean {
  return cssLangRE.test(request)
}

export function normalizePath(id: string): string {
  return path.posix.normalize(isWindows ? slash(id) : id)
}

const postfixRE = /[?#].*$/
export function cleanUrl(url: string): string {
  return url.replace(postfixRE, '')
}

export async function formatPostcssSourceMap(
  rawMap: ExistingRawSourceMap,
  file: string,
): Promise<ExistingRawSourceMap> {
  const inputFileDir = path.dirname(file)

  const sources = rawMap.sources.map((source) => {
    const cleanSource = cleanUrl(decodeURIComponent(source))

    // postcss virtual files
    if (cleanSource[0] === '<' && cleanSource.endsWith('>')) {
      return `\0${cleanSource}`
    }

    return normalizePath(path.resolve(inputFileDir, cleanSource))
  })

  return {
    file,
    mappings: rawMap.mappings,
    names: rawMap.names,
    sources,
    sourcesContent: rawMap.sourcesContent,
    version: rawMap.version,
  }
}
