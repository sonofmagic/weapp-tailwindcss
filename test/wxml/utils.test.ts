import { isPropsMatch } from '@/wxml/utils'

describe('utils', () => {
  it('isPropsMatch', () => {
    expect(isPropsMatch('a', 'a')).toBe(true)
  })
})
