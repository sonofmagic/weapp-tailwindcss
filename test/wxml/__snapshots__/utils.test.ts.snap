// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`utils > remove all id 1`] = `
"<view class="afterccontent-_m_ aftercml-0d5 afterctext-red-500 aspect-w-16" style="">
    <view class="itext-_h555_" style="">aspect</view>
    <view class="bg-_hfaf_" style="">w</view>
    <view class="bg-_h123_" style="">16</view>
  </view>"
`;

exports[`utils > remove all id case 0 1`] = `
"<view class="afterccontent-_m_ aftercml-0d5 afterctext-red-500 aspect-w-16" style="">
    <view class="itext-_h555_" style="">aspect</view>
    <view class="bg-_hfaf_" style="">w</view>
    <view class="bg-_h123_" style="">16</view>
  </view>"
`;
